File: sdk/dart/lib/src/wallet.dart
import 'dart:typed_data';
import 'package:hex/hex.dart';
import 'package:web3dart/web3dart.dart';
import 'package:bip39/bip39.dart' as bip39;
import 'package:http/http.dart';
import 'package:blockchain_sdk/src/constants.dart';
import 'dart:math';

/// A class representing an Ethereum wallet for blockchain interactions
class Wallet {
  /// The Ethereum credentials
  final EthPrivateKey _credentials;

  /// Web3 client for interacting with the blockchain
  final Web3Client _client;

  /// The wallet's Ethereum address
  String get address => _credentials.address.hex;

  /// The private key as a hex string
  String get privateKey => HEX.encode(_credentials.privateKey);

  /// Constructor with private key and RPC URL
  Wallet(String privateKey, String rpcUrl)
      : _credentials = EthPrivateKey.fromHex(privateKey),
        _client = Web3Client(rpcUrl, Client());

  /// Creates a new random wallet with connection to the specified RPC URL
  factory Wallet.createRandom(String rpcUrl) {
    final random = EthPrivateKey.createRandom(Random.secure());
    return Wallet(HEX.encode(random.privateKey), rpcUrl);
  }

  /// Creates a wallet from a mnemonic phrase
  factory Wallet.fromMnemonic(String mnemonic, String rpcUrl) {
    if (!bip39.validateMnemonic(mnemonic)) {
      throw Exception('Invalid mnemonic phrase');
    }

    final seed = bip39.mnemonicToSeed(mnemonic);
    // Simplified for now - in a real implementation, you would use a proper HD wallet derivation
    final privateKey = HEX.encode(seed.sublist(0, 32));
    return Wallet(privateKey, rpcUrl);
  }

  /// Generates a new random mnemonic phrase
  static String generateMnemonic() {
    return bip39.generateMnemonic(strength: 128); // 12 words
  }

  /// Get the ETH balance of the wallet
  Future<EtherAmount> getBalance() async {
    return await _client.getBalance(_credentials.address);
  }

  /// Send ETH to another address
  Future<String> sendTransaction({
    required String to,
    required BigInt amount,
    BigInt? gasPrice,
    BigInt? gasLimit,
    int? nonce,
  }) async {
    final transaction = Transaction(
      to: EthereumAddress.fromHex(to),
      value: EtherAmount.fromBigInt(EtherUnit.wei, amount),
      gasPrice: gasPrice != null
          ? EtherAmount.fromBigInt(EtherUnit.wei, gasPrice)
          : null,
      maxGas: gasLimit?.toInt() ?? BlockchainConstants.DEFAULT_GAS_LIMIT,
      nonce: nonce,
    );

    final txHash = await _client.sendTransaction(
      _credentials,
      transaction,
      chainId: null, // Chain ID should be specified for production
    );

    return txHash;
  }

  /// Sign a message with the private key
  Future<Uint8List> signMessage(Uint8List message) async {
    return _credentials.signPersonalMessageToUint8List(message);
  }

  /// Close the wallet and release resources
  void close() {
    _client.dispose();
  }
}

File: sdk/typescript/src/wallet.ts
import { ethers } from 'ethers';
import { SignedTransaction } from './types';

/**
 * Wallet for signing transactions
 */
export class Wallet {
  private signer: ethers.Wallet;

  /**
   * Create a wallet from a private key
   * @param privateKey Private key or mnemonic
   */
  constructor(privateKey: string) {
    this.signer = new ethers.Wallet(privateKey);
  }

  /**
   * Get wallet address
   */
  getAddress(): string {
    return this.signer.address;
  }

  /**
   * Sign a transaction
   * @param transaction Transaction object
   * @returns Signed transaction
   */
  async signTransaction(transaction: any): Promise<SignedTransaction> {
    // Convert to proper transaction format
    const tx: ethers.TransactionRequest = {
      to: transaction.to,
      data: transaction.data,
      value: transaction.value,
      gasLimit: transaction.gasLimit,
      gasPrice: transaction.gasPrice,
      nonce: transaction.nonce,
    };

    // Sign transaction
    const signedTx = await this.signer.signTransaction(tx);
    
    return {
      hash: ethers.keccak256(signedTx),
      from: this.signer.address,
      serialized: signedTx,
    };
  }

  /**
   * Sign a message
   * @param message Message to sign
   * @returns Signature
   */
  async signMessage(message: string | Uint8Array): Promise<string> {
    if (typeof message === 'string') {
      return this.signer.signMessage(message);
    } else {
      return this.signer.signMessage(message);
    }
  }

  /**
   * Create a random wallet
   * @returns New wallet instance
   */
  static createRandom(): Wallet {
    const randomWallet = ethers.Wallet.createRandom();
    return new Wallet(randomWallet.privateKey);
  }

  /**
   * Create a wallet from a mnemonic phrase
   * @param mnemonic Mnemonic phrase
   * @returns Wallet instance
   */
  static fromMnemonic(mnemonic: string): Wallet {
    const wallet = ethers.Wallet.fromPhrase(mnemonic);
    return new Wallet(wallet.privateKey);
  }

  /**
   * Create a wallet from a JSON keystore
   * @param json JSON keystore
   * @param password Password to decrypt
   * @returns Promise resolving to wallet
   */
  static async fromJson(json: string, password: string): Promise<Wallet> {
    const wallet = await ethers.Wallet.fromEncryptedJson(json, password);
    return new Wallet(wallet.privateKey);
  }

  /**
   * Export wallet to encrypted JSON
   * @param password Password to encrypt with
   * @returns Promise resolving to JSON string
   */
  async toJson(password: string): Promise<string> {
    return this.signer.encryptSync(password);
  }
} 